perf/reader_test.go

```diff
@@ -258,6 +260,305 @@ func TestPerfReaderLostSample(t *testing.T) {
 	}
 }
 
+func outputSamplesProgOverWritable(sampleSizes ...int) (*ebpf.Program, *ebpf.Map, error) {
+	const bpfFCurrentCPU = 0xffffffff
+
+	events, err := ebpf.NewMap(&ebpf.MapSpec{
+		Type: ebpf.PerfEventArray,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var maxSampleSize int
+	for _, sampleSize := range sampleSizes {
+		if sampleSize > maxSampleSize {
+			maxSampleSize = sampleSize
+		}
+	}
+
+	// Stash context somewhere
+	insns := asm.Instructions{
+		asm.LoadImm(asm.R0, 0, asm.DWord),
+		asm.Mov.Reg(asm.R9, asm.R1),
+	}
+
+	bufDwords := (maxSampleSize / 8) + 1
+	for i := 0; i < bufDwords; i++ {
+		insns = append(insns,
+			asm.StoreMem(asm.RFP, int16(i+1)*-8, asm.R0, asm.DWord),
+		)
+	}
+
+	for i, sampleSize := range sampleSizes {
+		insns = append(insns,
+			asm.Mov.Reg(asm.R1, asm.R9),
+			asm.LoadMapPtr(asm.R2, events.FD()),
+			asm.LoadImm(asm.R3, bpfFCurrentCPU, asm.DWord),
+			asm.LoadImm(asm.R0, int64(i), asm.DWord),
+			asm.StoreMem(asm.RFP, int16(bufDwords*-8), asm.R0, asm.DWord),
+			asm.Mov.Reg(asm.R4, asm.RFP),
+			asm.Add.Imm(asm.R4, int32(bufDwords*-8)),
+			asm.Mov.Imm(asm.R5, int32(sampleSize)),
+			asm.FnPerfEventOutput.Call(),
+		)
+	}
+
+	insns = append(insns, asm.Return())
+
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		License:      "GPL",
+		Type:         ebpf.XDP,
+		Instructions: insns,
+	})
+	if err != nil {
+		events.Close()
+		return nil, nil, err
+	}
+
+	return prog, events, nil
+}
+
+func mustOutputSamplesProgOverWritable(tb testing.TB, sampleSizes ...int) (*ebpf.Program, *ebpf.Map) {
+	tb.Helper()
+
+	// Requires at least 4.10 (9ecda41acb97 "perf/core: Add ::write_backward attribute to perf event")
+	testutils.SkipOnOldKernel(tb, "4.10", "overwritable perf events support")
+
+	prog, events, err := outputSamplesProgOverWritable(sampleSizes...)
+	var errVerifier *ebpf.VerifierError
+	if errors.As(err, &errVerifier) {
+		fmt.Printf("loading ebpf program:\n%+v", errVerifier)
+	}
+	if err != nil {
+		tb.Fatal(err)
+	}
+	tb.Cleanup(func() {
+		prog.Close()
+		events.Close()
+	})
+
+	return prog, events
+}
+
+func TestPerfReaderOverWritable(t *testing.T) {
```

Comment:
TestPerfReaderOverWritable and TestPerfReaderOverWritableOverWritten seem to be copies of TestPerfReaderLostSample, which is a really involved test because it needs to trigger a LostEvent record. As a result it's really hard to figure out what you are trying to test here.

The good news is that you don't need all that complexity: an overwritable perf ring will never generate a lost record by definition. Please rewrite your tests based on TestPerfReader instead. It might even be possible to test the new feature without outputSamplesProgOverWritable.